File: hotelInkaWEB/src/app/core/services/reserva/reserva.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from '../http/api.service';
import * as Url from '../../constants/api.constants';
import { Observable, tap } from 'rxjs';
import { ObtenerCatalogoHabitacionesDTO } from '../../models/reserva/response/lista/obtener-catalogo-habitaciones-dto.model';
import { ObtenerClientePorDNIDTO } from 'app/core/models/reserva/response/lista/obtener-cliente-por-dni-dto.model';
import { ObtenerReservaxDNI } from 'app/core/models/reserva/response/lista/obtener-reserva-por-dni-dto.model';
import { ObtenerOrdenHospedajeXDNI } from 'app/core/models/reserva/response/Obtener-orden-hospedaje-x-dni';
import { ObtenerCatalogoXTipoDTO } from 'app/core/models/reserva/response/Obtener-catalogo-servicios-por-tipo';
import { ObtenerTiposServiciosDTO } from 'app/core/models/reserva/response/lista/Obtener-tipo-servicio';
import { RegistrarOrdenServicio } from 'app/core/models/reserva/response/lista/Registrar-orden-servicio';
import { ResponseDTO } from 'app/core/models/generic/response-dto.model';

@Injectable({
  providedIn: 'root'
})
export class ReservaService {

  constructor(private apiService: ApiService,
  ) { }

  ObtenerCalogoHabitacionesAsync(): Observable<ObtenerCatalogoHabitacionesDTO[]> {
    return this.apiService.get(Url.Reserva.ObtenerCalogoHabitacionesAsync)
      .pipe(tap(data => data));
  }

  ObtenerClienteXDNIAsync(dni: string): Observable<ObtenerClientePorDNIDTO> {
    return this.apiService.query(Url.Reserva.ObtenerClienteXDNIAsync, { dni })
      .pipe(tap(data => data));
  }

  ObtenerReservaDNIAsync(idhueped: number): Observable<ObtenerReservaxDNI[]> {
    return this.apiService.query(Url.Reserva.ObtenerReservaDNIAsync, { idhueped })
      .pipe(tap(data => data));
  }

  ObtenerOrdenHospedajeAsync(dni: string): Observable<ObtenerOrdenHospedajeXDNI> {
    return this.apiService.query(Url.Reserva.ObtenerOrdenHospedajeAsync, { dni })
      .pipe(tap(data => data));
  }

  ObtenerCatalogoXTipoAsync(tipo: number): Observable<ObtenerCatalogoXTipoDTO[]> {
    return this.apiService.query(Url.Reserva.ObtenerCatalogoXTipoAsync, { tipo })
      .pipe(tap(data => data));
  }

  ObtenerTipoServicioAsync(): Observable<ObtenerTiposServiciosDTO[]> {
    return this.apiService.get(Url.Reserva.ObtenerTipoServicioAsync)
      .pipe(tap(data => data));
  }

  AddOrdenServicioAsync(request: RegistrarOrdenServicio): Observable<ResponseDTO> {
    return this.apiService.post(Url.Reserva.AddOrdenServicioAsync,request)
      .pipe(tap(data => data));
  }

}

File: Sol. HotelInka/HOTELINKA.APPLICATION/ReservaService.cs
using AutoMapper;
using HOTELINKA.DOMAIN.Domain;
using HOTELINKA.DOMAIN.Interface;
using HOTELINKA.DTO;
using HOTELINKA.DTO.Reserva.Request;
using HOTELINKA.DTO.Reserva.Response;

namespace HOTELINKA.APPLICATION
{
    public class ReservaService : IRservaService
    {
        private readonly IMapper _mapper;
        private readonly IReservaRepository _reservaRepository;

        public ReservaService(IMapper mapper,
                       IReservaRepository reservaRepository)
        {
            _mapper = mapper;
            _reservaRepository = reservaRepository;
        }

        public async Task<ResponseDTO> AddReservaAsync(RegistrarReservaRequest request)
        {
            var ultimoRegistroIdentificador = await _reservaRepository.GetLastCodigoReservaAsync();

            var newCodigo = ultimoRegistroIdentificador + 1;

            request.CodigoReserva = "C" + newCodigo.ToString().PadLeft(4, '0');

            return _mapper.Map<ResponseDTO>(await _reservaRepository.AddReservaAsync(_mapper.Map<Reserva>(request)));
        }
        public async Task<List<ObtenerCalogoHabitacionesDTO>> ObtenerCalogoHabitacionesAsync()
        {

            return _mapper.Map<List<ObtenerCalogoHabitacionesDTO>>(await _reservaRepository.GetCatalogoHabitaciones());
        }

        public async Task<List<ObtenerReservaxDNI>> ObtenerReservaDNIAsync(int idhuesped)
        {

            return _mapper.Map<List<ObtenerReservaxDNI>>(await _reservaRepository.GetReservasxCliente(idhuesped));
        }

        public async Task<ObtenerClientePorDNIDTO> ObtenerClienteXDNIAsync(string dni)
        {

            return _mapper.Map<ObtenerClientePorDNIDTO>(await _reservaRepository.GetClienteByDNI(dni));
        }

        public async Task<ResponseDTO> AddOrdenHospedajeAsync(RegistrarOrdenHospedaje request)
        {

            return _mapper.Map<ResponseDTO>(await _reservaRepository.AddOrdenHospedajeAsync(_mapper.Map<t01_orden_de_hospedaje>(request)));
        }
        //Obtener orden de hospedaje por x dni del cliente
        public async Task<ObtenerOrdenHospedajeXDNI> ObtenerOrdenHospedajeAsync(string DNI)
        {

            return _mapper.Map<ObtenerOrdenHospedajeXDNI>(await _reservaRepository.GetOrdenDeHospedajeXDNI(DNI));
        }

        public async Task<List<ObtenerCatalogoXTipoDTO>> ObtenerCatalogoXTipoAsync(int a)
        {

            return _mapper.Map<List<ObtenerCatalogoXTipoDTO>>(await _reservaRepository.GetObtenerCatalogoXTipo(a));
        }

        public async Task<List<ObtenerTiposServiciosDTO>> ObtenerTipoServicioAsync()
        {

            return _mapper.Map<List<ObtenerTiposServiciosDTO>>(await _reservaRepository.GetTipoServicio());
        }

        public async Task<ResponseDTO> AddOrdenServicioAsync(RegistrarOrdenServicio request)
        {

            return _mapper.Map<ResponseDTO>(await _reservaRepository.AddOrdenServicioAsync(_mapper.Map<t05_orden_de_servicios>(request)));
        }

        /*public async Task<ObtenerReservaDTO> ObtenerReservarxDNIAsync(string dni)
        {

            return _mapper.Map<ObtenerReservaDTO>(await _reservaRepository.GetReservaXDNI(dni));
        }*/
    }

}
